### ISSUE TITLE
proposal: `library_with_comments_or_annotations`

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/linter#3687]
## Description

Doc comments and/or annotations that are meant to be "scoped" to a library should be attached to a `library;` directive and not just "floating".

## Details

We have a bunch of silly rules trying to deal with doc comments and annotations "floating" that are meant to be scoped to a library (and not the first import/export directive).

Preferring these to be connected to a library directive would make their purpose clear and likely help formatting and code analysis tools.

## Kind

style

## Good Examples 

```dart
/// This is a cool library doc comment
@WithAnAnnotation()
library; // assuming https://github.com/dart-lang/language/issues/1073 lands with support for no-name library directives

import 'dart:async';
```

## Bad Examples

```dart
/// This is a bad floating library doc comment
@WithAFloatingAnnotation()

import 'dart:async';
```

## Discussion

This is the ONE case where https://github.com/dart-lang/linter/issues/3686 (prefer avoiding library directives) would not apply!

Adding no-name libraries proposal: https://github.com/dart-lang/language/issues/1073

This could likely be added to [Effective Dart], I think.

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn’t any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by srawlins** [comment]
The annotation bit might be tricky; it might be intentional. But the doc comment case I think is unambiguous.

**Comment by srawlins** [comment]
Perhaps if we only complain about annotations on the _first_ directive (like an import), our false negatives will be extremely minimal. And then not worry about annotations on the first member (like `main` or something).

**Comment by kevmoo** [comment]
> Perhaps if we only complain about annotations on the _first_ directive (like an import), our false negatives will be extremely minimal. And then not worry about annotations on the first member (like `main` or something).

Or (likely more difficult) lint when they are "floating" – with newlines below them.

**Comment by bwilkerson** [comment]
We could also flag for annotations whose target kind is 'library' because that's an unambiguous signal.

**Comment by kevmoo** [comment]
FYI: the associated language feature has LANDED for Dart 1.19!

**Comment by kevmoo** [comment]
Dropped `prefer_` prefix as is tradition

**Comment by srawlins** [comment]
Sorry I missed this request. These are each complete now, with the `library_annotations` and `dangling_library_doc_comments` lint rules.
